File: lab07/c++/p2.cpp
/**
 * Proiectarea Algoritmilor, 2016
 * Lab 7: Aplicatii DFS
 *
 * @author 	Radu Iacob
 * @email	radu.iacob23@gmail.com
 * @author adinu
 * @email  mandrei.dinu@gmail.com
 */

#include "graph.h"


/**
 * Calculeaza si afiseaza punctele de articulatie din graf
 *
 * Useful API:
 * graph.articulation_points
 *
 * Complexitate solutie: O(N + M)
 * N - numarul de noduri
 * M - numarul de muchii
 */

void dfs_articulation_points(Graph& g, Node* node, int father) {
	// TODO
}

void compute_articulation_points(Graph& g) {
	g.reset(); // reseteaza variabilele auxiliare

	// TODO

	std::cout << "Punctele de articulatie: \n";
	std::cout << g.articulation_points << "\n";
}

// **************************************************************//


/**
 * Calculeaza si afiseaza muchiile critice din graf
 *
 * Useful API:
 * graph.critical_edges
 *
 * Complexitate solutie: O(N + M)
 * N - numarul de noduri
 * M - numarul de muchii
 */

void dfs_critical_edges(Graph& g, Node* node, int father) {
	// TODO
}

void compute_critical_edges(Graph& g) {
	g.reset(); // reseteaza variabilele auxiliare

	// TODO
    std::cout << "Muchiile critice: \n";
	std::cout << g.critical_edges << "\n";
}

int main(int argc, char* argv[]) {

    std::fstream fin("test_undirected.in", std::fstream::in);

	int nr_teste;
	fin >> nr_teste;

	for (int i = 1; i <= nr_teste; ++i) {

	    std::cout << "TEST " << i << "\n";
		Graph g2(UNDIRECTED);
		fin >> g2;
		std::cout << g2;

		compute_articulation_points(g2);
		compute_critical_edges(g2);
		std::cout << "################################\n";
	}

	return 0;
}

File: lab07/java/src/P2.java
/**
 * Proiectarea Algoritmilor, 2016
 * Lab 7: Aplicatii DFS
 *
 * @author 	Radu Iacob
 * @email	radu.iacob23@gmail.com
 * @author adinu
 * @email  mandrei.dinu@gmail.com
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class P2 {

    static void DfsArticulationPoints(P1.Graph g, P1.Node node, int father) {
        // TODO
    }

    /**
     * Calculeaza si afiseaza punctele de articulatie din graf </br>
     * 
     * Useful API:                                            </br> 
     * List with the articulation points of the graph.        </br>
     * graph.articulationPoints                               </br>
     * 
     * Complexitate solutie: O(N + M) </br>
     * N - numarul de noduri          </br>
     * M - numarul de muchii          </br>
     * @param g
     */
    static void ArticulationPoints(P1.Graph g) {
        g.reset();

        // TODO
        // HINT: use dfsArticulationPoints after you implement it
      
       System.out.println("\nPuncte. de articulatie: \n" + g.articulationPoints);       
    }

    // ********************************************************************** //
        
    static void DfsCriticalEdges(P1.Graph g, P1.Node node, int father) {
        // TODO
    }

    /**
     * Calculeaza si afiseaza muchiile critice din graf. </br>
     * 
     * Useful API:
     * Lista cu muchiile critice      </br>
     * graph.criticalEdges            </br>
     *
     * Complexitate solutie: O(N + M) </br>
     * N - numarul de noduri          </br>
     * M - numarul de muchii          </br>
     * 
     * @param g
     */
    static void CriticalEdges(P1.Graph g) {
        g.reset();

        // TODO
        // Hint: use dfsCriticalEdges after you implement it

        System.out.println("\nMuchii critice: \n" + g.criticalEdges);
    }

    final static String PATH = "./res/test02";

    public static void main(String... args) throws FileNotFoundException {
        Scanner scanner = new Scanner(new File(PATH));

        int test_count = scanner.nextInt();
        for (int i = 1; i <= test_count; ++i) {
            System.out.println("TEST " + i + "\n");

            P1.Graph g = new P1.Graph(P1.Graph.GraphType.UNDIRECTED);
            g.readData(scanner);

            System.out.println(g);
            ArticulationPoints(g);
            CriticalEdges(g);
        }

        scanner.close();
    }
}
